### ISSUE TITLE
Huge v1.8 regression in sparse matrix division with diagonal matrix

### ISSUE THREAD

**Issue by wsshin** [JuliaSparse/SparseArrays.jl#290]
(The original issue was posted at https://github.com/JuliaLang/julia/issues/47673, but I was recommended to re-post it here.)

The two sparse matrices `A` and `B` are constructed by the finite difference method for partial differential equations.  `A` corresponds to a nearly diagonal matrix representing material parameters; `B` corresponds to a curl operator in a 2D Cartesian space.  Here is how they look like:
```julia-repl
julia> A
160000×160000 SparseMatrixCSC{ComplexF64, Int64} with 160000 stored entries:
⠑⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠑⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠑⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠑⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠑⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⢄⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⢄⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⢄⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⢄⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⢄

julia> B
160000×80000 SparseMatrixCSC{ComplexF64, Int64} with 320000 stored entries:
⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠈⢷⡀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⢷⡀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠈⢷⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀
⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧
```

Now, Here is the timing of first execution of `A \ B` in Julia 1.8.  It takes 557 seconds:
```julia-repl
julia> VERSION
v"1.8.3-pre.2"

julia> @timev A \ B
557.904421 seconds (7.08 M allocations: 364.345 MiB, 0.62% compilation time)
elapsed time (ns):  557904420654
gc time (ns):       0
bytes allocated:    382043682
pool allocs:        7070597
non-pool GC allocs: 4410
malloc() calls:     15
realloc() calls:    2
free() calls:       0
minor collections:  0
full collections:   0
160000×80000 SparseMatrixCSC{ComplexF64, Int64} with 320000 stored entries:
⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠈⢷⡀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⢷⡀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠈⢷⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀
⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧
```
Second execution is not much different:
```julia-repl
julia> @timev A \ B
557.779810 seconds (33 allocations: 19.246 MiB)
elapsed time (ns):  557779810182
gc time (ns):       0
bytes allocated:    20180968
pool allocs:        12
non-pool GC allocs: 13
malloc() calls:     6
realloc() calls:    2
free() calls:       0
minor collections:  0
full collections:   0
160000×80000 SparseMatrixCSC{ComplexF64, Int64} with 320000 stored entries:
⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠈⢷⡀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⢷⡀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠈⢷⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀
⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧
```

On the other hand, here is the timing of `A \ B` for the two matrices constructed by the same code, but on Julia 1.7.  First execution takes 3.2 seconds:
```julia-repl
julia> VERSION
v"1.7.2-pre.0"

julia> @timev A \ B
  3.199857 seconds (6.39 M allocations: 349.351 MiB, 2.93% gc time, 99.37% compilation time)
elapsed time (ns): 3199857054
gc time (ns):      93635079
bytes allocated:   366321440
pool allocs:       6385404
non-pool GC allocs:4326
malloc() calls:    7
realloc() calls:   2
GC pauses:         2
160000×80000 SparseMatrixCSC{ComplexF64, Int64} with 320000 stored entries:
⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠈⢷⡀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⢷⡀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠈⢷⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀
⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧
```
Second execution is even faster with 0.023 secs:
```julia-repl
julia> @timev A \ B
  0.023254 seconds (33 allocations: 19.246 MiB)
elapsed time (ns): 23254352
bytes allocated:   20180968
pool allocs:       12
non-pool GC allocs:13
malloc() calls:    6
realloc() calls:   2
160000×80000 SparseMatrixCSC{ComplexF64, Int64} with 320000 stored entries:
⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⢳⡀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠈⢷⡀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⢷⡀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠈⢷⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧⠀
⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢧
```

Is this a known issue?  Otherwise, please advise what I should do to further identify the cause of the regression.  The two matrices `A` and `B` are constructed by my big packages, so it is a bit difficult to generate a MWE demonstrating the regression without further guidance.

Here is the `versioninfo()` output for  Julia 1.8.
```julia-repl
julia> versioninfo()
Julia Version 1.8.3-pre.2
Commit 97ccb970c0 (2022-10-28 06:19 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin21.6.0)
  CPU: 8 × Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 1 on 4 virtual cores
Environment:
  JULIA_PKG_DEVDIR = /Users/wsshin/Programming
  JULIA_EDITOR = code
  JULIA_NUM_THREADS = 
```

**Comment by Wimmerer** [comment]
Can you serialize the matrices for me? And upload those files somewhere?

**Comment by SobhanMP** [comment]
```julia
using SparseArrays, LinearAlgebra

a = spzeros(ComplexF64, 100_000, 100_000) + I
b = spzeros(ComplexF64, 100_000, 50_000)
a \ b
```
no need to go far

**Comment by KristofferC** [comment]
In 1.7 the dispatch goes to:

https://github.com/JuliaLang/julia/blob/b06a24a7eed4464e50f4ee594c3a303ad543c42b/stdlib/LinearAlgebra/src/diagonal.jl#L595-L596

which dispatches to

https://github.com/JuliaSparse/SparseArrays.jl/blob/710fd7ba7184b062ca85e655e4b11928f6cc715d/src/linalg.jl#L892-L907

In 1.8 it first goes to

https://github.com/JuliaLang/julia/blob/0434deb161e17103eabdd7d82b17a1cd6b410572/stdlib/LinearAlgebra/src/diagonal.jl#L422-L423

but the first argument there does not evaluate to a diagonal:

```julia
2|julia> _init(\, D, B).(B) |> typeof
SparseMatrixCSC{ComplexF64, Int64}
```

so we dispatch to 

https://github.com/JuliaLang/julia/blob/0434deb161e17103eabdd7d82b17a1cd6b410572/stdlib/LinearAlgebra/src/diagonal.jl#L426-L440

which is not aware of sparsity.

**Comment by wsshin** [comment]
Is there a way to test the above merge myself?  I would like to confirm that the merge fixes my issue.  

I have tested merges for regular packages, but I'm not sure how to do it for standard libraries like `SparseArrays`.

**Comment by wsshin** [comment]
For example, if I build Julia from `HEAD` of the `master` branch, does it automatically retrieve `HEAD` of the `master` branch of all the standard libraries?

**Comment by Wimmerer** [comment]
No. What you can do is use Julia 1.9-DEV (or `master`) and then clone this repository. Change the `UUID` in `Project.toml` and then if you activate that `Project.toml` `using SparseArrays` will use that "local" version, rather than the built-in version.

**Comment by wsshin** [comment]
Can this be backported to Julia 1.8?  I thought v1.8.3 would be the final Julia 1.8, but I just saw that there is going to be v1.8.4: https://discourse.julialang.org/t/julia-1-8-4-testing-period/91755.  I am stuck to Julia 1.7 because of this issue.

**Comment by KristofferC** [comment]
I think it is backported

**Comment by wsshin** [comment]
I think we have this issue again in Julia 1.9.  Executing @SobhanMP's [example](https://github.com/JuliaSparse/SparseArrays.jl/issues/290#issuecomment-1324616181) in Julia 1.9:
```julia-repl
julia> VERSION
v"1.9.0-beta3.1"

julia> @time a \ b;
187.923855 seconds (2.63 M allocations: 184.404 MiB, 0.05% gc time, 1.45% compilation time)
```
On the other hand, in Julia 1.8:
```julia-repl
julia> VERSION
v"1.8.5"

julia> @time a \ b;
  0.005846 seconds (29 allocations: 7.390 MiB)
```

Can we create a unit test that checks this regression, so that the same bug does not keep reintroduced?

**Comment by dkarrasch** [comment]
> I think we have this issue again in Julia 1.9.

No, the fix hasn't been backported to v1.9 yet, but it is in the pipeline: https://github.com/JuliaSparse/SparseArrays.jl/pull/300

**Comment by wsshin** [comment]
I see.  Is there any prospect that it will be backported to 1.9 beta versions soon?  I started using 1.9 beta and noticed that FFTX was reduced a lot.  It is hard to go back to 1.8...

**Comment by wsshin** [comment]
For those of you who want to use this fix is Julia 1.9 betas:

- I tried to follow the steps that @Wimmerer [described](https://github.com/JuliaSparse/SparseArrays.jl/issues/290#issuecomment-1329323171);
- I tried to follow the steps shown in the [discourse material](https://discourse.julialang.org/t/testing-changes-to-stdlib-without-remaking-the-whole-of-julia/11998/9);
- I also tried to follow the steps mentioned in another [discourse material](https://discourse.julialang.org/t/edit-an-externally-hosted-standard-library/90740),

but none of these worked in a simple way.  

As a person who builds Julia from the source using `$ make`, the simplest way to build Julia 1.9 including this fix was as follows:

- Clone Julia repository:
```bash
$ git clone https://github.com/JuliaLang/julia.git
```

- Check out version 1.9:
```bash
$ cd julia
$ git checkout release-1.9
```

- The externally hosted standard libraries like SparseArrays seem to be checked out from GitHub during the Julia build procedure.  It turns out that the specific commits of these libraries to check out are described in `julia/stdlib/PKG_NAME.version`.  As of writing this posting, the contents of `julia/stdlib/SparseArrays.version` are

```
SPARSEARRAYS_BRANCH = main
SPARSEARRAYS_SHA1 = 4eaa4582569a76c3199849d8194582d948b7a70f
SPARSEARRAYS_GIT_URL := https://github.com/JuliaSparse/SparseArrays.jl.git
SPARSEARRAYS_TAR_URL = https://api.github.com/repos/JuliaSparse/SparseArrays.jl/tarball/$1
```

where `SPARSEARRAYS_SHA1` describes the SparseArrays commit that the Julia build procedure will use.  Change this line to

```
SPARSEARRAYS_SHA1 = a3ce57b90e8ffdf320d07506f9496ffcc4d421cb
```

which corresponds to the commit hash of the current fix shown at https://github.com/JuliaSparse/SparseArrays.jl/pull/291/commits.

- Now build Julia 1.9 by `$ make`.

This produced Julia 1.9 with the desired SparseArrays fix.  Hope this trick is useful for people like me.
